Failed to build OpenThread with the newer toolchain

It fails when building latest OpenThread with a newer toolchain, for example: [5-2016-q2-update](https://launchpad.net/gcc-arm-embedded/5.0/5-2016-q2-update):

Build steps:
```
./bootstrap
make -f example/Makefile-cc2538
```

Build failure logs:

```
In file included from /home/chenshu/code/openthread/examples/../src/core/openthread.cpp:47:0:
/home/chenshu/code/openthread/examples/../src/core/common/new.hpp:42:15: error: 'void operator delete(void*, size_t)' is a usual (non-placement) deallocation function in C++14 (or with -fsized-deallocation) [-Werror=c++14-compat]
 void OT_CDECL operator delete(void *, size_t) throw();
               ^
In file included from /home/chenshu/code/openthread/examples/../src/core/net/ip6.cpp:38:0:
/home/chenshu/code/openthread/examples/../src/core/common/new.hpp:42:15: error: 'void operator delete(void*, size_t)' is a usual (non-placement) deallocation function in C++14 (or with -fsized-deallocation) [-Werror=c++14-compat]
 void OT_CDECL operator delete(void *, size_t) throw();
               ^
/home/chenshu/code/openthread/examples/../src/core/openthread.cpp:66:15: error: 'void operator delete(void*, size_t)' is a usual (non-placement) deallocation function in C++14 (or with -fsized-deallocation) [-Werror=c++14-compat]
 void OT_CDECL operator delete(void *, size_t) throw() { }
```

We could build CC2538 platform with an older toolchain, for example: [4.9-2015-q3-update](https://launchpad.net/gcc-arm-embedded/4.9/4.9-2015-q3-update).

I still getting error while building on new toolchain with @wbober flag-addition tip. Below is the log.

Any inputs?
--------------------------------------------------------------------------------------------------------------------------------

Making all in apps
Making all in cli
  CCLD     ot-cli
../../../src/cli/libopenthread-cli.a(libopenthread_cli_a-cli_uart.o):(.rodata._ZTIN6Thread3Cli6ServerE[_ZTIN6Thread3Cli6ServerE]+0x0): undefined reference to `vtable for __cxxabiv1::__class_type_info'
../../../src/cli/libopenthread-cli.a(libopenthread_cli_a-cli_uart.o):(.rodata._ZTIN6Thread3Cli4UartE+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
/usr/bin/../lib/gcc/arm-none-eabi/5.4.1/armv7-m/libgcc.a(unwind-arm.o): In function `get_eit_entry':
unwind-arm.c:(.text+0x138): undefined reference to `__exidx_start'
unwind-arm.c:(.text+0x13c): undefined reference to `__exidx_end'
collect2: error: ld returned 1 exit status
Makefile:478: recipe for target 'ot-cli' failed
make[5]: *** [ot-cli] Error 1
Makefile:422: recipe for target 'all-recursive' failed
make[4]: *** [all-recursive] Error 1
Makefile:424: recipe for target 'all-recursive' failed
make[3]: *** [all-recursive] Error 1
Makefile:578: recipe for target 'all-recursive' failed
make[2]: *** [all-recursive] Error 1
Makefile:503: recipe for target 'all' failed
make[1]: *** [all] Error 2
examples/Makefile-cc2538:235: recipe for target 'do-build-cortex-m3' failed
make: *** [do-build-cortex-m3] Error 2

--- a/.travis.yml
+++ b/.travis.yml
@@ -48,7 +48,10 @@ matrix:
     - env: BUILD_TARGET="scan-build"
       compiler: clang
       os: linux
-    - env: BUILD_TARGET="cc2538"
+    - env: BUILD_TARGET="arm-gcc49"
+      compiler: gcc
+      os: linux
+    - env: BUILD_TARGET="arm-gcc54"
       compiler: gcc
       os: linux
     - env: BUILD_TARGET="posix"

--- a/.travis/before_install.sh
+++ b/.travis/before_install.sh
@@ -67,11 +67,19 @@ cd /tmp || die
         sudo apt-get install clang || die
     }
 
-    [ $BUILD_TARGET != cc2538 ] || {
+    [ $BUILD_TARGET != arm-gcc49 ] || {
         sudo apt-get install lib32z1 || die
         wget https://launchpad.net/gcc-arm-embedded/4.9/4.9-2015-q3-update/+download/gcc-arm-none-eabi-4_9-2015q3-20150921-linux.tar.bz2 || die
         tar xjf gcc-arm-none-eabi-4_9-2015q3-20150921-linux.tar.bz2 || die
         export PATH=/tmp/gcc-arm-none-eabi-4_9-2015q3/bin:$PATH || die
+        arm-none-eabi-gcc --version || die
+    }
+
+    [ $BUILD_TARGET != arm-gcc54 ] || {
+        sudo apt-get install lib32z1 || die
+        wget https://launchpad.net/gcc-arm-embedded/5.0/5-2016-q3-update/+download/gcc-arm-none-eabi-5_4-2016q3-20160926-linux.tar.bz2 || die
+        tar xjf gcc-arm-none-eabi-5_4-2016q3-20160926-linux.tar.bz2 || die
+        export PATH=/tmp/gcc-arm-none-eabi-5_4-2016q3/bin:$PATH || die
         arm-none-eabi-gcc --version || die
     }
 

--- a/.travis/script.sh
+++ b/.travis/script.sh
@@ -47,8 +47,15 @@ set -x
     scan-build --status-bugs -analyze-headers -v make || die
 }
 
-[ $BUILD_TARGET != cc2538 ] || {
+[ $BUILD_TARGET != arm-gcc49 ] || {
     export PATH=/tmp/gcc-arm-none-eabi-4_9-2015q3/bin:$PATH || die
+    COMMISSIONER=1 JOINER=1 make -f examples/Makefile-cc2538 || die
+    arm-none-eabi-size  output/bin/arm-none-eabi-ot-cli || die
+    arm-none-eabi-size  output/bin/arm-none-eabi-ot-ncp || die
+}
+
+[ $BUILD_TARGET != arm-gcc54 ] || {
+    export PATH=/tmp/gcc-arm-none-eabi-5_4-2016q3/bin:$PATH || die
     COMMISSIONER=1 JOINER=1 make -f examples/Makefile-cc2538 || die
     arm-none-eabi-size  output/bin/arm-none-eabi-ot-cli || die
     arm-none-eabi-size  output/bin/arm-none-eabi-ot-ncp || die

--- a/configure.ac
+++ b/configure.ac
@@ -224,7 +224,7 @@ AC_PROG_LN_S
 #
 
 PROSPECTIVE_CFLAGS="-Wall -Wextra -Wshadow -Werror -std=c99 -pedantic-errors"
-PROSPECTIVE_CXXFLAGS="-Wall -Wextra -Wshadow -Werror"
+PROSPECTIVE_CXXFLAGS="-Wall -Wextra -Wshadow -Werror -Wno-c++14-compat"
 
 AC_CACHE_CHECK([whether $CC is Clang],
     [nl_cv_clang],

